typed-racket-lib/typed-racket/rep/type-rep.rkt

```diff
@@ -449,6 +451,51 @@
      (define d* (remove-duplicates d))
      (if (and (pair? d*) (null? (cdr d*))) (car d*) d*))])
 
+
+;; Intersection
+(def-type Intersection ([elems (and/c (set/c Type/c)
+                                      (λ (s) (>= (set-count s) 2)))])
+  [#:intern (for/set ([e (in-set elems)])
+              (Rep-seq e))]
+  [#:frees (λ (f) (combine-frees (for/list ([elem (in-set elems)])
+                                   (f elem))))]
+  [#:fold-rhs (let ([elems (for/list ([elem (in-set elems)])
+                             (type-rec-id elem))])
+                (apply -intersect elems))]
+  [#:key (let ()
+           (define d
+             (let loop ([ts (set->list elems)] [res null])
+               (cond [(null? ts) res]
+                     [else
+                      (define k (Type-key (car ts)))
+                      (cond [(not k) (list #f)]
+                            [(pair? k) (loop (cdr ts) (append k res))]
+                            [else (loop (cdr ts) (cons k res))])])))
+           (define d* (remove-duplicates d))
+           (if (and (pair? d*) (null? (cdr d*))) (car d*) d*))])
+
+
+(define (-intersect . ts)
+  (let loop ([elems (list)]
+             [ts ts])
+    (match ts
+      [(list)
+       (define t-set (list->set elems))
+       (cond
+         [(set-empty? t-set) (Univ)]
+         [(= 1 (set-count t-set)) (set-first t-set)]
+         [else (*Intersection t-set)])]
```

Comment:
The set creation should go in the else clause, the other two can be tested on `elems`.